File: __tests__/theBox/gen_test.ml

open Jest
module Gen = TheBox.OfGenerators
let (forAll) = Quickie.Property.(forAll)
let (fixSize) = Quickie.Generator.(fixSize)

module Quick = Quickie.Make(struct
  type t = unit case
  let ok () = Just Ok
  let fail s = Just (Fail s)
end)

let _ =

test "int < size" (fun _ ->
  Quick.check
    (forAll
      Gen.int
      (fun n -> Gen.int (fun () -> n) n < n)
    )
);

test "int > 0" (fun _ ->
  Quick.check
    (forAll
      Gen.int
      (fun n -> n > 0)
    )
);

test {js|int - 10±5% in bottom decile|js} (fun _ ->
  let max = (1 lsl 30) - 1 in
  let sampleSize = 1000 in
  let quantileSizePercentage = 10 in
  let quantileSize = sampleSize / quantileSizePercentage in
  let acceptedDeviation = (quantileSize / 2) in
  let quantileThreshold = max / quantileSizePercentage in
  let quantile v = v |> Js.Array.filter ((>) quantileThreshold) in
  let deviation q = q |> Array.length |> ((-) quantileSize) |> abs in
  Quick.check
    (forAll
      (fixSize sampleSize (Gen.arrayOf (fixSize max Gen.int)))
      (fun v ->
        v |> quantile
          |> deviation
          |> ((>=) acceptedDeviation)
      )
      ~message: (fun v _ ->
        let actual = (v |> quantile |> deviation |> string_of_int) in
        let accepted = (string_of_int acceptedDeviation) in
        "Sample deviates by " ^ actual ^ " which is greater than the accepted deviation of " ^ accepted
      )
    )
);

test {js|int - 10±5% in top decile|js} (fun _ ->
  let max = (1 lsl 30) - 1 in
  let sampleSize = 1000 in
  let quantileDivisor = 10 in
  let quantileSize = sampleSize / quantileDivisor in
  let acceptedDeviation = (quantileSize / 2) in
  let quantileThreshold = max - (max / quantileDivisor) in
  let quantile v = v |> Js.Array.filter ((<) quantileThreshold) in
  let deviation q = q |> Array.length |> ((-) quantileSize) |> abs in
  Quick.check
    (forAll
      (fixSize sampleSize (Gen.arrayOf (fixSize max Gen.int)))
      (fun v ->
        v |> quantile
          |> deviation
          |> ((>=) acceptedDeviation)
      )
      ~message: (fun v _ ->
        let actual = (v |> quantile |> deviation |> string_of_int) in
        let accepted = (string_of_int acceptedDeviation) in
        "Sample deviates by " ^ actual ^ " which is greater than the accepted deviation of " ^ accepted
      )
    )
);

test {js|choose - 10±5% in bottom decile|js} (fun _ ->
  let max = 100 in
  let sampleSize = 1000 in
  let quantileDivisor = 10 in
  let quantileSize = sampleSize / quantileDivisor in
  let acceptedDeviation = (quantileSize / 2) in
  let quantileThreshold = max / quantileDivisor in
  let quantile v = v |> Js.Array.filter ((>) quantileThreshold) in
  let deviation q = q |> Array.length |> ((-) quantileSize) |> abs in
  Quick.check
    (forAll
      (fixSize sampleSize (Gen.arrayOf (fixSize max (Gen.choose 0 max))))
      (fun v ->
        v |> quantile
          |> deviation
          |> ((>=) acceptedDeviation)
      )
      ~message: (fun v _ ->
        let actual = (v |> quantile |> deviation |> string_of_int) in
        let accepted = (string_of_int acceptedDeviation) in
        "Sample deviates by " ^ actual ^ " which is greater than the accepted deviation of " ^ accepted
      )
    )
);

test {js|choose - 10±5% in top decile|js} (fun _ ->
  let max = 4600 in
  let sampleSize = 1000 in
  let quantileDivisor = 10 in
  let quantileSize = sampleSize / quantileDivisor in
  let acceptedDeviation = (quantileSize / 2) in
  let quantileThreshold = max - (max / quantileDivisor) in
  let quantile v = v |> Js.Array.filter ((<) quantileThreshold) in
  let deviation q = q |> Array.length |> ((-) quantileSize) |> abs in
  Quick.check
    (forAll
      (fixSize sampleSize (Gen.arrayOf (fixSize max (Gen.choose 0 max))))
      (fun v ->
        v |> quantile
          |> deviation
          |> ((>=) acceptedDeviation)
      )
      ~message: (fun v _ ->
        let actual = (v |> quantile |> deviation |> string_of_int) in
        let accepted = (string_of_int acceptedDeviation) in
        "Sample deviates by " ^ actual ^ " which is greater than the accepted deviation of " ^ accepted
      )
    )
);

test {js|oneOf - ±10% even distribution|js} (fun _ ->
  let choices = [| "a"; "b"; "c" |] in
  let sampleSize = 1000 in
  let target = sampleSize / Array.length choices in
  let acceptedDeviation = target / 10 in
  let count arr value =
    arr |> Js.Array.filter ((=) value)
        |> Array.length
    in
  Quick.check
    (forAll
      (fixSize sampleSize (Gen.arrayOf (Gen.oneOf choices)))
      (fun arr ->
        choices
          |> Array.map (count arr)
          |> Array.map ((-) target)
          |> Array.map abs
          |> Js.Array.filter ((>) acceptedDeviation)
          |> Array.length
          |> ((<) 0)
      )
      ~message: (fun v _ ->
        let target = string_of_int target in
        let accepted = string_of_int acceptedDeviation in
        let distribution =
          choices
            |> Array.map (fun c -> (c, count v c))
            |> Array.map (fun (c, n) -> c ^ ": " ^ (string_of_int n))
            |> Js.Array.joinWith "\r\n\t"
          in
        "Sample is too unevenly distributed (target: " ^ target ^ {js|±|js} ^ accepted ^ "):\r\n\t" ^ distribution
      )
    )
);

File: lib/js/__tests__/theBox/gen_test.js
'use strict';

var Jest       = require("bs-jest/lib/js/src/jest.js");
var $$Array    = require("bs-platform/lib/js/array.js");
var Curry      = require("bs-platform/lib/js/curry.js");
var TheBox     = require("../../src/theBox.js");
var Quickie    = require("../../src/quickie.js");
var Caml_obj   = require("bs-platform/lib/js/caml_obj.js");
var Caml_int32 = require("bs-platform/lib/js/caml_int32.js");
var Pervasives = require("bs-platform/lib/js/pervasives.js");

var forAll = Quickie.Property[/* forAll */0];

var fixSize = Quickie.Generator[/* fixSize */0];

function ok() {
  return Jest.pass;
}

var fail = Jest.fail;

var Quick = Quickie.Make(/* module */[
      /* ok */ok,
      /* fail */fail
    ]);

Jest.test("int < size", (function () {
        return Curry._3(Quick[/* check */0], /* None */0, /* None */0, Curry._3(forAll, /* None */0, TheBox.OfGenerators[/* int */0], (function (n) {
                          return +(Curry._2(TheBox.OfGenerators[/* int */0], (function () {
                                          return n;
                                        }), n) < n);
                        })));
      }));

Jest.test("int > 0", (function () {
        return Curry._3(Quick[/* check */0], /* None */0, /* None */0, Curry._3(forAll, /* None */0, TheBox.OfGenerators[/* int */0], (function (n) {
                          return +(n > 0);
                        })));
      }));

Jest.test("int - 10±5% in bottom decile", (function () {
        var quantileSize = 100;
        var acceptedDeviation = quantileSize / 2 | 0;
        var quantileThreshold = 107374182;
        var quantile = function (v) {
          return v.filter((function (param) {
                        return Caml_obj.caml_greaterthan(quantileThreshold, param);
                      }));
        };
        var deviation = function (q) {
          return Pervasives.abs(quantileSize - q.length | 0);
        };
        return Curry._3(Quick[/* check */0], /* None */0, /* None */0, Curry._3(forAll, /* Some */[(function (v, _) {
                            var actual = Pervasives.string_of_int(deviation(quantile(v)));
                            var accepted = Pervasives.string_of_int(acceptedDeviation);
                            return "Sample deviates by " + (actual + (" which is greater than the accepted deviation of " + accepted));
                          })], Curry._2(fixSize, 1000, Curry._1(TheBox.OfGenerators[/* arrayOf */5], Curry._2(fixSize, 1073741823, TheBox.OfGenerators[/* int */0]))), (function (v) {
                          return Caml_obj.caml_greaterequal(acceptedDeviation, deviation(quantile(v)));
                        })));
      }));

Jest.test("int - 10±5% in top decile", (function () {
        var quantileSize = 100;
        var acceptedDeviation = quantileSize / 2 | 0;
        var quantileThreshold = 966367641;
        var quantile = function (v) {
          return v.filter((function (param) {
                        return Caml_obj.caml_lessthan(quantileThreshold, param);
                      }));
        };
        var deviation = function (q) {
          return Pervasives.abs(quantileSize - q.length | 0);
        };
        return Curry._3(Quick[/* check */0], /* None */0, /* None */0, Curry._3(forAll, /* Some */[(function (v, _) {
                            var actual = Pervasives.string_of_int(deviation(quantile(v)));
                            var accepted = Pervasives.string_of_int(acceptedDeviation);
                            return "Sample deviates by " + (actual + (" which is greater than the accepted deviation of " + accepted));
                          })], Curry._2(fixSize, 1000, Curry._1(TheBox.OfGenerators[/* arrayOf */5], Curry._2(fixSize, 1073741823, TheBox.OfGenerators[/* int */0]))), (function (v) {
                          return Caml_obj.caml_greaterequal(acceptedDeviation, deviation(quantile(v)));
                        })));
      }));

Jest.test("choose - 10±5% in bottom decile", (function () {
        var quantileSize = 100;
        var acceptedDeviation = quantileSize / 2 | 0;
        var quantileThreshold = 10;
        var quantile = function (v) {
          return v.filter((function (param) {
                        return Caml_obj.caml_greaterthan(quantileThreshold, param);
                      }));
        };
        var deviation = function (q) {
          return Pervasives.abs(quantileSize - q.length | 0);
        };
        return Curry._3(Quick[/* check */0], /* None */0, /* None */0, Curry._3(forAll, /* Some */[(function (v, _) {
                            var actual = Pervasives.string_of_int(deviation(quantile(v)));
                            var accepted = Pervasives.string_of_int(acceptedDeviation);
                            return "Sample deviates by " + (actual + (" which is greater than the accepted deviation of " + accepted));
                          })], Curry._2(fixSize, 1000, Curry._1(TheBox.OfGenerators[/* arrayOf */5], Curry._2(fixSize, 100, Curry._2(TheBox.OfGenerators[/* choose */1], 0, 100)))), (function (v) {
                          return Caml_obj.caml_greaterequal(acceptedDeviation, deviation(quantile(v)));
                        })));
      }));

Jest.test("choose - 10±5% in top decile", (function () {
        var quantileSize = 100;
        var acceptedDeviation = quantileSize / 2 | 0;
        var quantileThreshold = 4140;
        var quantile = function (v) {
          return v.filter((function (param) {
                        return Caml_obj.caml_lessthan(quantileThreshold, param);
                      }));
        };
        var deviation = function (q) {
          return Pervasives.abs(quantileSize - q.length | 0);
        };
        return Curry._3(Quick[/* check */0], /* None */0, /* None */0, Curry._3(forAll, /* Some */[(function (v, _) {
                            var actual = Pervasives.string_of_int(deviation(quantile(v)));
                            var accepted = Pervasives.string_of_int(acceptedDeviation);
                            return "Sample deviates by " + (actual + (" which is greater than the accepted deviation of " + accepted));
                          })], Curry._2(fixSize, 1000, Curry._1(TheBox.OfGenerators[/* arrayOf */5], Curry._2(fixSize, 4600, Curry._2(TheBox.OfGenerators[/* choose */1], 0, 4600)))), (function (v) {
                          return Caml_obj.caml_greaterequal(acceptedDeviation, deviation(quantile(v)));
                        })));
      }));

Jest.test("oneOf - ±10% even distribution", (function () {
        var choices = /* array */[
          "a",
          "b",
          "c"
        ];
        var target = Caml_int32.div(1000, choices.length);
        var acceptedDeviation = target / 10 | 0;
        var count = function (arr, value) {
          return arr.filter((function (param) {
                        return Caml_obj.caml_equal(value, param);
                      })).length;
        };
        return Curry._3(Quick[/* check */0], /* None */0, /* None */0, Curry._3(forAll, /* Some */[(function (v, _) {
                            var target$1 = Pervasives.string_of_int(target);
                            var accepted = Pervasives.string_of_int(acceptedDeviation);
                            var distribution = $$Array.map((function (param) {
                                      return param[0] + (": " + Pervasives.string_of_int(param[1]));
                                    }), $$Array.map((function (c) {
                                          return /* tuple */[
                                                  c,
                                                  count(v, c)
                                                ];
                                        }), choices)).join("\r\n\t");
                            return "Sample is too unevenly distributed (target: " + (target$1 + ("±" + (accepted + ("):\r\n\t" + distribution))));
                          })], Curry._2(fixSize, 1000, Curry._1(TheBox.OfGenerators[/* arrayOf */5], Curry._1(TheBox.OfGenerators[/* oneOf */2], $$Array.map((function (x, _, _$1) {
                                          return x;
                                        }), choices)))), (function (arr) {
                          return Caml_obj.caml_lessthan(0, $$Array.map(Pervasives.abs, $$Array.map((function (param) {
                                                  return target - param | 0;
                                                }), $$Array.map((function (param) {
                                                      return count(arr, param);
                                                    }), choices))).filter((function (param) {
                                            return Caml_obj.caml_greaterthan(acceptedDeviation, param);
                                          })).length);
                        })));
      }));

var Gen = 0;

exports.forAll  = forAll;
exports.fixSize = fixSize;
exports.Gen     = Gen;
exports.Quick   = Quick;
/* Quick Not a pure module */
